flink-runtime/src/test/java/org/apache/flink/runtime/util/profiler/ProfilingServiceTest.java

```diff
@@ -0,0 +1,170 @@
+    @Test
+    void testFailedRequestUnderProfiling() throws ExecutionException, InterruptedException {
+        ProfilingInfo profilingInfo =
+                profilingService
+                        .requestProfiling(resourceID, 10, ProfilingInfo.ProfilingMode.ITIMER)
+                        .get();
+        Assertions.assertEquals(ProfilingInfo.ProfilingStatus.RUNNING, profilingInfo.getStatus());
+        try {
+            profilingService
+                    .requestProfiling(
+                            resourceID, profilingDuration, ProfilingInfo.ProfilingMode.ITIMER)
+                    .get();
+            Assertions.fail("Duplicate profiling request should throw with IllegalStateException.");
+        } catch (Exception e) {
+            Assertions.assertTrue(e.getCause() instanceof IllegalStateException);
+        }
+    }
+
+    @Test
+    @Timeout(value = 1, unit = TimeUnit.MINUTES)
+    public void testAllProfilingMode() throws ExecutionException, InterruptedException {
+        for (ProfilingInfo.ProfilingMode mode : ProfilingInfo.ProfilingMode.values()) {
+            ProfilingInfo profilingInfo =
+                    profilingService.requestProfiling(resourceID, profilingDuration, mode).get();
+            if (isNoPermissionOrAllocateSymbol(profilingInfo)) {
+                LOG.warn(
+                        "Ignoring failed profiling instance in {} mode, which caused by no permission.",
+                        profilingInfo.getProfilingMode());
+                continue;
+            }
+            Assertions.assertEquals(
+                    ProfilingInfo.ProfilingStatus.RUNNING,
+                    profilingInfo.getStatus(),
+                    String.format(
+                            "Submitting profiling request should be succeed or no permission, but got errorMsg=%s",
+                            profilingInfo.getMessage()));
+            waitForProfilingFinished(profilingService);
+            Assertions.assertEquals(
+                    ProfilingInfo.ProfilingStatus.FINISHED,
+                    profilingInfo.getStatus(),
+                    String.format(
+                            "Profiling request should complete successful, but got errorMsg=%s",
+                            profilingInfo.getMessage()));
+        }
+
+        verifyRollingDeletion(profilingService);
+    }
+
+    private void verifyConfigsWorks(ProfilingService profilingService, Path configuredDir) {
+        Assertions.assertEquals(configuredDir.toString(), profilingService.getProfilingResultDir());
+        Assertions.assertEquals(historySizeLimit, profilingService.getHistorySizeLimit());
+    }
+
+    private void verifyRollingDeletion(ProfilingService profilingService) {
+        ArrayDeque<ProfilingInfo> profilingList =
+                profilingService.getProfilingListForTest(resourceID);
+        // Profiling History shouldn't exceed history size limit.
+        Assertions.assertEquals(historySizeLimit, profilingList.size());
```

Comment:
Can we always assert this if some tests cannot be taken due to lacking permission.
How about `Assertions.assertTrue(profilingList.size() <= profilingService.getHistorySizeLimit());`?

Reply:
It's safe here because there are five profiling modes(only `CPU` and `ALLOCATION` are affected by permission or AllocTracer symbols) in the async-profiler. But it makes sense to replace them with `<=`. Corrected.